### PR TITLE
Allow nested with multiple nesting and subcolumns of complex types

### LINKED ISSUE #18826: Cannot access named tuple element by its name using dot
**Describe the bug**
Consider a column `a` of type `Tuple(s String, i Int64)`. Currently we can access certain field by using any of the following expressions:

```
tupleElement(a, 1)
a.1
tupleElement(a, 's')
```

but what seems to be the most convenient way for named tuples does not work:

```
a.s
```

It would be nice to support accessing tuple elements by their names.

Reproduction on current Arcadia version from revision r7736838 (which is similar to github master around Jan 7th):

```
max42-dev.sas.yp-c.yandex.net :) create table t(a Tuple(s String, i Int8)) engine = Log()

CREATE TABLE t
(
    `a` Tuple(s String, i Int8)
)
ENGINE = Log

Query id: f9cbbcf2-0db4-4312-bd7c-ee3e901bfdd7

Ok.

0 rows in set. Elapsed: 0.004 sec.

max42-dev.sas.yp-c.yandex.net :) insert into t values (('x', 42))

INSERT INTO t VALUES

Query id: acd443b2-1daa-43c3-b201-8713b0a2250a

Ok.

1 rows in set. Elapsed: 0.008 sec.

max42-dev.sas.yp-c.yandex.net :) select a from t

SELECT a
FROM t

Query id: bb38d164-29b6-49f4-a35c-df9f5b0b7d42

┌─a────────┐
│ ('x',42) │
└──────────┘

1 rows in set. Elapsed: 0.005 sec.

max42-dev.sas.yp-c.yandex.net :) select a.s from t

SELECT a.s
FROM t

Query id: de107034-0831-409d-a7a3-81a60c81a360


Received exception from server (version 20.13.1):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Missing columns: 'a.s' while processing query: 'SELECT a.s FROM t', required columns: 'a.s', maybe you meant: .

0 rows in set. Elapsed: 0.003 sec.

max42-dev.sas.yp-c.yandex.net :) select tupleElement(a, 's') from t

SELECT tupleElement(a, 's')
FROM t

Query id: b5454a30-1c16-4b8f-a4e1-9d1fde5bdf9b

┌─tupleElement(a, 's')─┐
│ x                    │
└──────────────────────┘

1 rows in set. Elapsed: 0.006 sec.

max42-dev.sas.yp-c.yandex.net :) select a.1 from t

SELECT a.1
FROM t

Query id: 29442b78-4235-410e-bfb1-6b6c0f624fa5

┌─tupleElement(a, 1)─┐
│ x                  │
└────────────────────┘

1 rows in set. Elapsed: 0.006 sec.
```

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Data type `Nested` now supports arbitrary levels of nesting. Introduced subcolumns of complex types, such as `size0` in `Array`, `null` in `Nullable`, names of `Tuple` elements, which can be read without reading of whole column.

Detailed description / Documentation draft:
Continuation of #14963.

This closes #18826.

### CODE REVIEW

File: src/DataTypes/DataTypeOneElementTuple.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+#include <Columns/IColumn.h>
+
+
+namespace DB
+{
+
+namespace
+{
+
+class DataTypeOneElementTupleStreams : public IDataTypeCustomStreams
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeOneElementTuple.cpp:17]: Missing comments.

File: src/DataTypes/IDataType.h [hunk 2]
```diff
@@ -101,20 +103,31 @@ class IDataType : private boost::noncopyable
         /// Index of tuple element, starting at 1.
         String tuple_element_name;
 
+        bool escape_tuple_delimiter = true;
```
Comment (alexey-milovidov) [inline src/DataTypes/IDataType.h:106]: Missing comment.

File: src/DataTypes/IDataType.h [hunk 3]
```diff
@@ -101,20 +103,31 @@ class IDataType : private boost::noncopyable
...
+
         Substream(Type type_) : type(type_) {}
+
+        String toString() const;
+    };
+
+    struct SubstreamPath : public std::vector<Substream>
```
Comment (alexey-milovidov) [inline src/DataTypes/IDataType.h:113]: .

File: src/DataTypes/IDataType.h [hunk 4]
```diff
@@ -101,20 +103,31 @@ class IDataType : private boost::noncopyable
...
+    {
+        String toString() const;
     };
 
-    using SubstreamPath = std::vector<Substream>;
+    using SubstreamsCache = std::unordered_map<String, ColumnPtr>;
```
Comment (alexey-milovidov) [inline src/DataTypes/IDataType.h:118]: .

File: src/DataTypes/IDataType.h [hunk 5]
```diff
@@ -458,9 +503,11 @@ class IDataType : private boost::noncopyable
     /// This is mutable to allow setting custom name and serialization on `const IDataType` post construction.
     mutable DataTypeCustomNamePtr custom_name;
     mutable DataTypeCustomTextSerializationPtr custom_text_serialization;
+    mutable DataTypeCustomStreamsPtr custom_streams;
```
Comment (alexey-milovidov) [inline src/DataTypes/IDataType.h:506]: Need to provide motivating example in comments.

Comment (CurtizJ) [reply src/DataTypes/IDataType.h:509]: I've added comment to `IDataTypeCustomStreams`.

File: src/Storages/MergeTree/MergeTreeDataPartWide.cpp [hunk 6]
```diff
@@ -46,10 +48,13 @@ IMergeTreeDataPart::MergeTreeReaderPtr MergeTreeDataPartWide::getReader(
     const ValueSizeMap & avg_value_size_hints,
     const ReadBufferFromFileBase::ProfileCallback & profile_callback) const
 {
+    auto new_settings = reader_settings;
+    new_settings.convert_nested_to_subcolumns = true;
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/MergeTreeDataPartWide.cpp:52]: Don't understand.

File: src/DataTypes/DataTypeArray.cpp [hunk 7]
```diff
@@ -145,10 +148,48 @@ namespace
...
+        if (column_offsets.empty())
+            return column_sizes;
+
+        const auto & offsets_data = column_offsets.getData();
+        auto & sizes_data = assert_cast<ColumnArray::ColumnOffsets &>(*column_sizes).getData();
+
+        sizes_data.resize(offsets_data.size());
+        sizes_data[0] = offsets_data[0];
+        for (size_t i = 1; i < offsets_data.size(); ++i)
+            sizes_data[i] = offsets_data[i] - offsets_data[i - 1];
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeArray.cpp:166]: I expect that compiler will generate suboptimal code due to possible aliasing.

File: src/DataTypes/DataTypeArray.cpp [hunk 8]
```diff
@@ -145,10 +148,48 @@ namespace
...
+        if (column_sizes.empty())
+            return column_offsets;
+
+        const auto & sizes_data = column_sizes.getData();
+        auto & offsets_data = assert_cast<ColumnArray::ColumnOffsets &>(*column_offsets).getData();
+
+        offsets_data.resize(sizes_data.size());
+        offsets_data[0] = sizes_data[0];
+        for (size_t i = 0; i < sizes_data.size(); ++i)
+            offsets_data[i] = offsets_data[i - 1] + sizes_data[i];
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeArray.cpp:185]: The same issue as above.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Looks like conversion from ANTLR AST is wrong. I will disable this test for ANTLR.

**Pull request merged**